README.md
## Overview
------------

PowerGym is a Gym-like environment for Volt-Var control in power distribution systems.

The Volt-Var control targets minimizing voltage violations, control loss, and power loss under physical networked constraints and device constraints. The networked constraints are maintained by the power distribution system simulator, OpenDSS. The device constraints are usually integer constraints on the actions.

Below is a description of observation and action spaces. {} denotes a finite set and [] denote a continuous interval.

|**Observation Space** | |
| ------------- | ------------- |
| **Variable**| **Range**|
| Bus voltage     | [0.8, 1.2] |
| Capacitor status     | {0, 1} |
| Regulator tap number | {0, ..., 32} |
| State-of-charge (soc) | [0, 1] |
| Discharge power  | [-1, 1]  |

|**Action Space** | |
| ------------- | ------------- |
| **Variable**| **Range**|
| Capacitor status     | {0, 1} |
| Regulator tap number | {0, ..., 32} |
| Discharge power (disc.) | {0, ..., 32} |
| Discharge power (cont.) | [-1, 1]  |

There are two kinds of batteries. Discrete battery has discretized choices on the discharge power (e.g., choose from {0,...,32}) and continuous battery chooses the normalized discharge power from the interval [-1,1]. The user should specify the battery's kind upon calling the environment.

The reward function is a combination of three losses: voltage violation, control error, and power loss. The control error is further decomposed into capacitor's & regulator's switching cost and battery's discharge loss & soc loss. The weights among these losses depends on the circuit system and is listed in the Appendix of our paper. 

The implemented circuit systems are summerized as follows.
| **System**| **# Caps**| **# Regs**| **# Bats**|
| ------------- | ------------- |------- |------- |
| 13Bus     | 2 | 3 | 1 |
| 34Bus | 4 | 6 | 2 |
| 123Bus | 4 | 7 | 4 |
| 8500Node | 10 | 12 | 10 |


## Requirements
------------
- Python 3.8 or higher

For the complete installation
```
pip install -r requirements.txt
```

## Usage
------------
### Run options
`random_agent.py` gives a minimal example of PowerGym usage. The option `--mode` can choose various running mode

To run PowerGym in a single episode
 ```
 python random_agent.py
 ```

To run PowerGym for parallel environments
```
python random_agent.py --mode=parallele
```

To run PowerGym for multiple episodes
```
python random_agent.py --mode=episodic
```

To run PowerGym using OpenDSS controllers defined in the circuit files (if any) 
```
python random_agent.py --mode=dss
```

### Environment name options
The option `--env_name` can choose various environments. Below, we take 123Bus as an example.

Run a vanilla environment
```
python random_agent.py --env_name 123Bus
```

Run a scaled environment
```
python random_agent.py --env_name 123Bus_s1.5
```

Run an environment with soc error
```
python random_agent.py --env_name 123Bus_soc
```

Run a scaled environment with soc error
```
python random_agent.py --env_name 123Bus_soc_s1.5
```



## Citation

To cite PowerGym, please cite the following paper:

```
@article{fan2021powergym,
  title={PowerGym: A Reinforcement Learning Environment for Volt-Var Control in Power Distribution Systems},
  author={Fan, Ting-Han and Lee, Xian Yeow and Wang, Yubo},
  journal={arXiv preprint arXiv:2109.03970},
  year={2021}
}
```

## License
This project is licensed under MIT License. See [LICENSE.md](LICENSE.md) for more details.


CHANGELOG.md
- Initial Release


powergym/loadprofile.py
# Copyright 2021 Siemens Corporation
# SPDX-License-Identifier: MIT

import numpy as np
import pandas as pd
import os
from pathlib import Path
from fnmatch import fnmatch

class LoadProfile:
    def __init__(self, steps, dss_folder_path, dss_file, worker_idx=None):
        self.steps = steps
        
        self.dss_folder_path = dss_folder_path
        self.loadshape_path = os.path.join(dss_folder_path, 'loadshape')
        if worker_idx is None:
            self.loadshape_dss = 'loadshape.dss'
        else:
            self.loadshape_dss = 'loadshape_' + str(worker_idx) + '.dss'
        
        self.LOAD_NAMES = self.find_load_names(dss_file)
    
        self.FILES = []
        for f in os.listdir(self.loadshape_path):
            low = f.lower()
            if ('loadshape' in low) and low.endswith('.csv'):
                self.FILES.append(os.path.join(self.loadshape_path, f))
        

    def create_file_with_daily(self, fname):
        '''
            Create a new file named fname[:-4]_daily.dss
            In the new file,
            if any load is created, then the load is associated with its daily loadshape.
        '''
        fin = open(os.path.join(self.dss_folder_path, fname), 'r')
        fout = open(os.path.join(self.dss_folder_path, fname[:-4] + '_daily.dss'), 'w')
        for line in fin:
            if not line.lower().startswith('new load.') or ('daily' in line):
                fout.write(line)
            else:
                line = line.strip()
                if '!' in line: line = line[:line.find('!')].strip() # remove inline comment
                if '//' in line: line = line[:line.find('//')].strip() # remove inline comment
                spt = list(filter(None, line.split(' '))) # filter out the empty string
                load = spt[1].split('.',1)[1]
                fout.write(line + ' daily=loadshape_' + load + '\n')
        fin.close()
        fout.close()
   
    def add_redirect_and_mode_at_main_daily_dss(self, main_daily_dss):
        '''
            Add redirect loadshape (& load file if any) 
            and set daily mode at the main daily dss file

        Args:
            main_daily_dss: the file name of the main daily dss file
        
        Returns:
            the load dss file (if any) associated with the main dss file
        '''
        # load the file
        fin = open(os.path.join(self.dss_folder_path, main_daily_dss), 'r')
        lines = [line for line in fin]
        fin.close()

        # overwrite the file
        found_load, redirect_load = False, False
        load_file = None
        fout = open(os.path.join(self.dss_folder_path, main_daily_dss), 'w')
        for line in lines:
            low = line.strip().lower()
            if '!' in low: low = low[:low.find('!')].strip() # remove inline comment
            if '//' in low: low = low[:low.find('//')].strip() # remove inline comment
            if (not found_load) and 'load' in low and not low.startswith('~'):
                fout.write('! add loadshape\n')
                fout.write('redirect ' + self.loadshape_dss + '\n\n')
                found_load = True

            low = low[:-4] if len(low)>=4 else ''
            if (not redirect_load) and low.startswith('redirect'):
                if low.endswith('loads') or low.endswith('load'):
                    load_file = list(filter(None, line.strip().split(' ')))[1] # remove the empty string
                    fout.write('redirect ' + load_file[:-4] + '_daily.dss\n')
                    redirect_load = True
                elif low.endswith('loads_daily') or low.endswith('load_daily'):
                    load_file = list(filter(None, line.strip().split(' ')))[1] # remove the empty string
                    fout.write(line)
                    redirect_load = True
                else: fout.write(line)
            else: fout.write(line)
        
        assert found_load, 'cannot find load at ' + main_daily_dss

        fout.write('Set mode=Daily number=1 hour=0 stepsize=3600 sec=0\n')
        fout.close()
        
        return load_file

    def find_load_file_from(self, main_dss):
        fin = open(os.path.join(self.dss_folder_path, main_dss), 'r')
        load_file = None
        for line in fin:
            low = line.strip().lower()
            if '!' in low: low = low[:low.find('!')].strip() # remove inline comment
            if '//' in low: low = low[:low.find('//')].strip() # remove inline comment
            low = low[:-4] if len(low)>=4 else ''
            if low.startswith('redirect') and \
               (low.endswith('loads') or low.endswith('load') \
                or low.endswith('loads_daily') or low.endswith('load_daily') ):
                
                load_file = list(filter(None, line.strip().split(' ')))[1]
                break
        return load_file

    def find_load_names(self, main_dss):
        '''
            Find the loads with daily loadshapes at main dss or the load dss files.
            If there is none, 
            then generate new files (annotated _daily) with daily loadshapes.
        '''
        def find_load_name(fname, names):
            file_path = os.path.join(self.dss_folder_path, fname)
            assert os.path.exists(file_path), file_path + ' not found'
            
            needs_load_daily, daily_mode = False, False
            with open(file_path, 'r') as fin:
                for line in fin:
                    low = line.strip().lower()
                    if low.startswith('new load.'):
                        if 'daily' in low:
                            spt = line.split(' ')
                            spt = list(filter(None, spt)) # filter out the empty string
                            names.append(spt[1].split('.',1)[1])
                        else: needs_load_daily = True
                    if low.startswith('set mode=daily'):
                        daily_mode = True
            return needs_load_daily, daily_mode
        names = []

        # add from the main dss file
        needs_load_daily, daily_mode = find_load_name(main_dss, names)
        if needs_load_daily or (not daily_mode):
            ## Create a new _daily file. Add daily loadshape if needed
            self.create_file_with_daily(main_dss)

            ## add redirect and set daily mode at the new _daily file
            load_file = self.add_redirect_and_mode_at_main_daily_dss(\
                                              main_dss[:-4]+'_daily.dss')
        else:
            load_file = self.find_load_file_from(main_dss)

        # add from the other load files
        if load_file is not None:
            if 'daily' in load_file:
                needs_load_daily, _ = find_load_name(load_file, names)
                assert (not needs_load_daily), 'invalid content in ' + load_file
            else:
                needs_load_daily, _ = find_load_name(load_file, names)
                if needs_load_daily: self.create_file_with_daily(load_file)

        # check empty or duplicate load
        assert len(names)>0, 'daliy load not found. Consider modifying from the auto-generated file annotated with _daily'
        assert len(names) == len(set(names)), 'duplicate load names'
        
        return names     

    def gen_loadprofile(self, scale=1.0):
        
        try:
            dfs = []
            for f in self.FILES:
                dfs.append( pd.read_csv(f, header=None) )
            assert len(dfs)>0, r'put load shapes files under ./loadshape'
            df = pd.concat(dfs).rename(columns = {0: 'mul'}).reset_index(drop=True)
            if scale!=1.0: df['mul'] = df['mul']*scale
        except:
            print(r'put load shapes files under ./loadshape')
        
        # Since we've concatenated all loadshape file together,
        # the first load uses the top steps*days data points
        # the second load uses the second chunk and so on.
        
        # compute totoal number of episodes
        episodes = len(df) // ( self.steps * len(self.LOAD_NAMES) )
        
        # stop here only if the loadprofile folders exist
        checks = [fnmatch(f, '0*') for f in os.listdir(self.loadshape_path)]
        scale_txt = os.path.join(self.loadshape_path, 'scale.txt')
        fscale = np.genfromtxt(scale_txt).reshape(1)[0] if os.path.exists(scale_txt) else None
        if sum(checks)==episodes and fscale==scale:
            return episodes

        # save the scale for future use
        np.savetxt(scale_txt, np.array([scale]))

        # insert loadname, day, step columns
        load_col, episode_col, step_col = [], [], []
        for i in range( self.steps*episodes*len(self.LOAD_NAMES) ):
            load_col.append(self.LOAD_NAMES[i//(self.steps*episodes)])
            episode_col.append((i//self.steps)%episodes)
            step_col.append(i%self.steps)
        df = df[:len(load_col)]
        df['load'] = load_col
        df['episode'] = episode_col
        df['step'] = step_col
        
        # sort and output
        df = df.sort_values(by=['episode','load','step'])[['episode','load','step','mul']].reset_index(drop=True)
        for episode in range(episodes):
            if not os.path.exists(os.path.join(self.loadshape_path, str(episode).zfill(3))):
                os.mkdir(os.path.join(self.loadshape_path, str(episode).zfill(3)))
            sdf = df[df['episode']==episode]
            for load in self.LOAD_NAMES:
                series = sdf[sdf['load']==load]['mul']
                series.to_csv(\
                  os.path.join(self.loadshape_path, str(episode).zfill(3), load+'.csv'),\
                  header=False, index=False)

        return episodes # number of distinct epochs
    
    def choose_loadprofile(self, idx):
        assert os.path.exists(os.path.join(self.loadshape_path, str(idx).zfill(3))), 'idx does not exist'
        
        with open(os.path.join(self.dss_folder_path, self.loadshape_dss), 'w') as fp:
            for load_name in self.LOAD_NAMES:
                fp.write(f'New Loadshape.loadshape_{load_name} npts={self.steps} sinterval={60*60*24//self.steps} ' +
                    'mult=(file=./' + os.path.join('loadshape', str(idx).zfill(3), load_name+'.csv') + ')\n' )

        return os.path.join(self.loadshape_path, str(idx).zfill(3))

    def get_loadprofile(self, idx):
        folder_path  = os.path.join(self.loadshape_path, str(idx).zfill(3))
        csv_paths = os.listdir(folder_path)
        temp_loads = []
        for csv in csv_paths:
            csv_file = os.path.join(folder_path,csv)
            load = pd.read_csv(csv_file, header=None, names=[csv.split('.')[0]])
            temp_loads.append(load)

        all_loads = pd.concat(temp_loads, axis=1)
        return all_loads


powergym/env_register.py
# Copyright 2021 Siemens Corporation
# SPDX-License-Identifier: MIT

import os
import inspect
import re
from powergym.env import Env

# map from system_name to fixed information of the system

_SYS_INFO = {
    '13Bus': {
        'source_bus': 'sourcebus',
        'node_size': 500,
        'shift': 10,
        'show_node_labels': True
    },

    '34Bus': {
        'source_bus': 'sourcebus',
        'node_size': 500,
        'shift': 80,
        'show_node_labels': True
    },

    '123Bus': {
        'source_bus': '150',
        'node_size': 400,
        'shift': 80,
        'show_node_labels': True
    },

    '8500-Node': {
        'source_bus': 'e192860',
        'node_size': 10,
        'shift': 0,
        'show_node_labels': False
    }
}


# map from env_name to the necessary information
_ENV_INFO = {
    '13Bus': {
        'system_name': '13Bus',
        'dss_file': 'IEEE13Nodeckt_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': 33,
        'power_w': 10.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 0.0/33,
        'dis_w': 6.0/33
    },
   
    '13Bus_cbat': {
        'system_name': '13Bus',
        'dss_file': 'IEEE13Nodeckt_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': float('inf'),
        'power_w': 10.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 0.0/33,
        'dis_w': 6.0/33,
    },
   
    '13Bus_soc': {
        'system_name': '13Bus',
        'dss_file': 'IEEE13Nodeckt_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': 33,
        'power_w': 10.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 20.0/33,
        'dis_w': 1.0/33
    },

    '13Bus_cbat_soc': {
        'system_name': '13Bus',
        'dss_file': 'IEEE13Nodeckt_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': float('inf'),
        'power_w': 10.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 20.0/33,
        'dis_w': 1.0/33,
    },

    '34Bus': {
        'system_name': '34Bus',
        'dss_file': 'ieee34Mod1_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': 33,
        'power_w': 1.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 0.0/33,
        'dis_w': 10.0/33,
    },

    '34Bus_cbat': {
        'system_name': '34Bus',
        'dss_file': 'ieee34Mod1_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': float('inf'),
        'power_w': 1.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 0.0/33,
        'dis_w': 10.0/33,
    },

    '34Bus_soc': {
        'system_name': '34Bus',
        'dss_file': 'ieee34Mod1_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': 33,
        'power_w': 1.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 500.0/33,
        'dis_w': 4.0/33,
    },

    '34Bus_cbat_soc': {
        'system_name': '34Bus',
        'dss_file': 'ieee34Mod1_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': float('inf'),
        'power_w': 1.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 500.0/33,
        'dis_w': 4.0/33,
    },

    '123Bus': {
        'system_name': '123Bus',
        'dss_file': 'IEEE123Master_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': 33,
        'power_w': 10.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 0.0/33,
        'dis_w': 7.0/33,
    },

    '123Bus_cbat': {
        'system_name': '123Bus',
        'dss_file': 'IEEE123Master_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': float('inf'),
        'power_w': 10.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 0.0/33,
        'dis_w': 7.0/33,
    },

    '123Bus_soc': {
        'system_name': '123Bus',
        'dss_file': 'IEEE123Master_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': 33,
        'power_w': 10.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 500.0/33,
        'dis_w': 5.0/33,
    },

    '123Bus_cbat_soc': {
        'system_name': '123Bus',
        'dss_file': 'IEEE123Master_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': float('inf'),
        'power_w': 10.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 500.0/33,
        'dis_w': 5.0/33,
    },

    '8500Node': {
        'system_name': '8500-Node',
        'dss_file': 'Master_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': 33,
        'power_w': 1.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 0.0/33,
        'dis_w': 200.0/33,
    },
    
    '8500Node_cbat': {
        'system_name': '8500-Node',
        'dss_file': 'Master_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': float('inf'),
        'power_w': 1.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 0.0/33,
        'dis_w': 200.0/33,
    },


    '8500Node_soc': {
        'system_name': '8500-Node',
        'dss_file': 'Master_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': 33,
        'power_w': 1.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 10000/33,
        'dis_w': 100/33,
    },

    '8500Node_cbat_soc': {
        'system_name': '8500-Node',
        'dss_file': 'Master_daily.dss',
        'max_episode_steps': 24,
        'reg_act_num': 33,
        'bat_act_num': float('inf'),
        'power_w': 1.0,
        'cap_w': 1.0/33,
        'reg_w': 1.0/33,
        'soc_w': 10000/33,
        'dis_w': 100/33,
    }
}

# add system information to environment
for env in _ENV_INFO.keys():
    sys = _ENV_INFO[env]['system_name']
    _ENV_INFO[env].update(_SYS_INFO[sys])



####################### functions ########################
def get_info_and_folder(env_name):
    # check env scale and env name
    is_scaled = re.match('.*(_s)([0-9]*[.])?[0-9]+?', env_name)
    if is_scaled:
        matched_str = is_scaled.group(0)
        idx = matched_str.rfind('_s')
        env_name = matched_str[:idx]
        scale = float(matched_str[idx+2:])
    assert env_name in _ENV_INFO, env_name + ' not implemented'

    # get base_info
    base_info = _ENV_INFO[env_name].copy()
    if is_scaled:
        base_info['scale'] = scale
        base_info['soc_w'] = base_info['soc_w'] * (scale**2)

    # get folder path
    folder_path = os.path.join(os.path.dirname(os.path.abspath(inspect.getsourcefile(Env))), '..', 'systems')
    folder_path = os.path.abspath(folder_path)
    return base_info, folder_path

def make_env(env_name, dss_act=False, worker_idx=None):
    base_info, folder_path = get_info_and_folder(env_name)

    if worker_idx is None:
        return Env(folder_path, base_info, dss_act)
    else:
        base_file = os.path.join(folder_path, base_info['system_name'], base_info['dss_file'])
        assert os.path.exists(base_file), base_file + ' does not exist'
        fin = open(base_file, 'r')
        
        with open(base_file[:-4] + '_' + str(worker_idx) + '.dss', 'w') as fout:
            for line in fin:
                if line.strip() == 'redirect loadshape.dss':
                    fout.write('redirect loadshape_' + str(worker_idx) + '.dss\n')
                else:
                    fout.write(line)
        info = base_info.copy()
        info['dss_file'] = info['dss_file'][:-4] + '_' + str(worker_idx) + '.dss'
        info['worker_idx'] = worker_idx
        return Env(folder_path, info, dss_act)
        
def remove_parallel_dss(env_name, num_workers):
    base_info, folder_path = get_info_and_folder(env_name)
    base_main = os.path.join(folder_path, base_info['system_name'], base_info['dss_file'])
    base_loadshape = os.path.join(folder_path, base_info['system_name'], 'loadshape.dss')

    bases = [base_main, base_loadshape]
    for base in bases:
        for i in range(num_workers):
            fname = base[:-4] + '_' + str(i) + '.dss'
            if os.path.exists(fname):
                os.remove(fname)


random_agent.py
# Copyright 2021 Siemens Corporation
# SPDX-License-Identifier: MIT

"""Random agent to probe enviroment
"""
import matplotlib.pyplot as plt
import numpy as np
import imageio
import glob

from powergym.env_register import make_env, remove_parallel_dss

import argparse
import random
import itertools
import sys, os
import multiprocessing as mp

def parse_arguments():
    parser = argparse.ArgumentParser(description='Argument Parser')
    parser.add_argument('--env_name', default='13Bus')
    parser.add_argument('--seed', type=int, default=123456, metavar='N',
                         help='random seed')
    parser.add_argument('--num_steps', type=int, default=1000, metavar='N',
                         help='maximum number of steps')
    parser.add_argument('--num_workers', type=int, default=3, metavar='N', 
                         help='number of parallel processes')
    parser.add_argument('--use_plot', type=lambda x: str(x).lower()=='true', default=False)
    parser.add_argument('--do_testing', type=lambda x: str(x).lower()=='true', default=False)
    parser.add_argument('--mode', type=str, default='single',
                        help="running mode, random, parallele, episodic or dss")
    args = parser.parse_args()
    return args

def seeding(seed):
    np.random.seed(seed)
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)


def run_random_agent(args, load_profile_idx=0, worker_idx=None, use_plot=False, print_step=False):
    """Run a agent that takes arbitrary control actions
    
    Args:
        test_action (str): Test env with fixed or random actions 
        steps (int, optional): Number of steps to run the env
        use_plot (bool, optional): Plot outcome of random agent
    """

    cwd = os.getcwd()
    
    # get environment
    env = make_env(args.env_name, worker_idx=worker_idx)
    env.seed(args.seed + 0 if worker_idx is None else worker_idx)

    if print_step:
        print('This system has {} capacitors, {} regulators and {} batteries'.format(env.cap_num, env.reg_num, env.bat_num))
        print('reg, bat action nums: ', env.reg_act_num, env.bat_act_num)
        print('-'*80)

    # assuming we only train on one load profile for now
    obs = env.reset(load_profile_idx=load_profile_idx)
    
    # RL loop
    if use_plot and not os.path.exists(os.path.join(cwd,'random_agent_plots')):
        os.makedirs(os.path.join(cwd,'random_agent_plots'))

    episode_reward = 0.0
    for i in range(env.horizon):
        action = env.random_action()
        obs, reward, done, info = env.step(action)
        episode_reward += reward
        print(worker_idx, i, reward)

        if print_step:
            print('\nStep:{}\n'.format(i))
            print('Action:', action)
            print('Next Obs: {} R: {} Done: {} Info: {}'.format(obs, reward, done, info))

        if use_plot:
            # node_bound argument to decide to plot max or min node voltage for nodes with more than 1 phase
            fig, _ = env.plot_graph()
            fig.tight_layout(pad=0.1)
            fig.savefig(os.path.join(cwd,'random_agent_plots', 'node_voltage_' + str(i).zfill(4) + '.png'))
            plt.close()
    
    print('load_profile: {}, episode_reward: {}'.format(load_profile_idx, episode_reward))

    if use_plot:
        fig, _ = env.plot_graph(show_voltages=False)
        fig.tight_layout(pad=0.1)
        fig.savefig(os.path.join(cwd, 'random_agent_plots', 'system_layout.pdf'))
    
        # make sure to import imageio and glob if generating gif
        # generate gif
        '''
        images = []
        filenames = sorted(glob.glob(os.path.join(cwd, "random_agent_plots/*.png")))
        for filename in filenames:
            images.append(imageio.imread(filename))
        imageio.mimsave(os.path.join(cwd, 'random_agent_plots/node_voltage.gif'), images, fps=1)
        '''

def run_parallel_random_agent(args):
    workers = []
    for i in range(0, args.num_workers):
        p = mp.Process( target=run_random_agent, args=(args,i, i) )
        p.start()
        workers.append(p)
    for p in workers:
        p.join()

    remove_parallel_dss(args.env_name, args.num_workers)

def random_evaluate(env, profiles, episodes=None):
    returns = np.zeros(len(profiles)) if episodes is None else np.zeros(min(episodes, len(profiles)))
    for i in range(len(returns)):
        pidx = profiles[i] if episodes is None else random.choice(profiles)
        obs = env.reset(load_profile_idx = pidx)
        episode_reward = 0
        episode_steps = 0
        done = False
        while not done:
            action = env.random_action()
            next_obs, reward, done, _ = env.step(action)
            episode_reward += reward
            episode_steps += 1
            mask = 1 if episode_steps == env.horizon else float(not done)
            obs = next_obs
        returns[i] = episode_reward
    return returns.mean(), returns.std()

def run_episodic_random_agent(args, worker_idx=None):
    """Run a episodic random agent with train-test split
  
    """
    # output file
    if args.do_testing:
        fout = open("result/{}_random_{}.csv".format(args.env_name, args.seed), 'w')

    # get environment
    env = make_env(args.env_name, worker_idx=worker_idx)
    env.seed(args.seed + 0 if worker_idx is None else worker_idx)

    # get obs, act
    obs_dim = env.observation_space.shape[0]
    CRB_num = ( env.cap_num, env.reg_num, env.bat_num )
    CRB_dim = (2, env.reg_act_num, env.bat_act_num )
    print('NumCap, NumReg, NumBat: {}'.format(CRB_num))
    print('ObsDim, ActDim: {}, {}'.format(obs_dim, sum(CRB_num)))
    print('-'*80)

    # train-test split
    if args.do_testing:
        train_profiles = random.sample(range(env.num_profiles), k=env.num_profiles//2)
        test_profiles = [i for i in range(env.num_profiles) if i not in train_profiles]
    else:
        train_profiles = list(range(env.num_profiles))

    # Training Loop
    total_numsteps = 0
        
    for i_episode in itertools.count(start=1):
        
        episode_reward = 0
        episode_steps = 0
        done = False
        load_profile_idx = random.choice(train_profiles)
        obs = env.reset(load_profile_idx = load_profile_idx)

        while not done:
            action = env.random_action()  # Sample random action
            next_obs, reward, done, info = env.step(action)
            episode_steps += 1
            total_numsteps += 1
            episode_reward += reward
            mask = 1 if episode_steps == env.horizon else float(not done)
            obs = next_obs

        print("episode: {}, profile: {}, total numsteps: {}, episode steps: {}, reward: {}".format(i_episode, load_profile_idx, total_numsteps, episode_steps, round(episode_reward, 2)))
        
        total_numsteps+=24

        if total_numsteps >= args.num_steps: break
        
        if args.do_testing and i_episode%5==0:
            avg_reward, std = random_evaluate(env, test_profiles)
            fout.write('{},{},{}\n'.format(total_numsteps, avg_reward, std))
            fout.flush()
            print("----------------------------------------")
            print("Avg., Std. Reward: {}, {}".format(round(avg_reward, 2), round(std, 2)))
            print("----------------------------------------")

def run_dss_agent(args):  
    # get environment
    env = make_env(args.env_name, dss_act=True)
    env.seed(args.seed)
    
    profiles = list(range(env.num_profiles))

    # Training Loop
    total_numsteps = 0
        
    for i_episode in itertools.count(start=1):
        episode_reward = 0
        episode_steps = 0
        done = False
        load_profile_idx = random.choice(profiles)
        obs = env.reset(load_profile_idx = load_profile_idx)

        while not done:
            next_obs, reward, done, info = env.dss_step()  
            episode_steps += 1
            total_numsteps += 1
            episode_reward += reward
            obs = next_obs
            
        print("episode: {}, profile: {}, total numsteps: {}, episode steps: {}, reward: {}".format(i_episode, load_profile_idx, total_numsteps, episode_steps, round(episode_reward, 2))) 
        if total_numsteps >= args.num_steps: break

if __name__ == '__main__':
    args = parse_arguments()
    seeding(args.seed)
    if args.mode.lower() ==  'single':
        run_random_agent(args, worker_idx=None, use_plot=args.use_plot, print_step=False)
    elif args.mode.lower() ==  'parallele':
        run_parallel_random_agent(args)
    elif args.mode.lower() == 'episodic':
        run_episodic_random_agent(args)
    elif args.mode.lower() == 'dss':
        run_dss_agent(args)
    else:
        raise NotImplementedError("Running mode not implemented")
